### PR TITLE
Bug 1163120 - Zombies

### PR DESCRIPTION
**Pull request by wesj**
This lets a Browser exist without a webView. The webView within it can be created/destroyed by calling zombify()/resurrect(). While its a zombie, it holds on to (some) information that's given to it so that it can handle it later when its resurrected. i.e. the last load request. A navigation delegate. I started to make userScripts that were loaded work like that as well, but decided it was easier for callers to just be careful about re-registering themselves when the tabs were resurrected. There's a delegate interface for notifications when tabs are zombified or resurrected.

Tabs are automatically "resurrected" when they're created (unless you specify not to), or when they're selected. Zombie tabs load the thumbnails from thumbnail storage. 

I also... couldn't resist and made a few other little changes. This completely zombifies all unselected tabs on memory warnings. Seems like something we should do? It also removes selecting tabs from the addTab method, since the two seem like orthogonal things to me. i.e. Ideally you'd just call:
tabManager.selectedTab = tabManager.addTab(url)

### CODE REVIEW

**Comment by thebnich** [comment]
A few questions/comments before I review:

Zombifying tabs on low memory does seem like something we should do -- but it's already done for us. That was the cause of [bug 1145402](https://bugzilla.mozilla.org/show_bug.cgi?id=1145402).

With that in mind, it might be reasonable for us to drop `zombify()/resurrect()` entirely, letting the OS handle zombification, and letting the fix in [bug 1145402](https://bugzilla.mozilla.org/show_bug.cgi?id=1145402) handle resurrection. The only thing we really need is a way for the webview to be lazily loaded at startup. Could this all be simplified to making `Browser`'s webView a lazy var?

**Comment by wesj** [comment]
I started down that path originally, but we access the webView in didCreateTab, which basically makes the laziness worthless. As soon as you start trying to move away from that you basically end up here. In fact, there's no code in BrowserViewController.didCreateTab anymore. I thought about removing it, but left it in for now. didAddTab actually does touch the webView as well I think. Those are all optional-? accesses now, but with lazy vars, they'd also wind up having to know about the state of the tab.

**Comment by wesj** [comment]
I'm not really clear what the webview does under memory pressure. i.e. the webview itself seems to exist, but has to be reloaded, right? I don't mind dropping that bit here. Was looking for an easy way to test this (firing a memory pressure event manually).

**Comment by thebnich** [comment]
Yeah, the webview instance doesn't die, just the actual page content. I found that it kills only the number of webviews needed to get out of the low memory state, rather than killing them all at once, which was kind of neat.

For whatever reason, I found that the OS zombification doesn't happen when the app is connected to the Xcode debugger.

**Comment by thebnich** [comment]
> [...] but with lazy vars, they'd also wind up having to know about the state of the tab.

Yeah, I didn't literally mean `lazy` vars, just webViews that are lazily loaded. That is, we could create the Browser/webView which holds its restore URL, but only once it's actually selected (similar to [bug 1145402](https://bugzilla.mozilla.org/show_bug.cgi?id=1145402)). Will that work?

**Comment by wesj** [comment]
There's some questions there about the overhead of creating a webview but not loading anything in it that I really don't know much about. But we'd essentially end up with resurrect/zombify methods in that case that flipped some boolean to switch modes? What do we gain at that point?

**Comment by thebnich** [comment]
> But we'd essentially end up with resurrect/zombify methods in that case that flipped some boolean to switch modes?

Not sure what you mean -- we wouldn't have "zombify" at all (since it's done by the OS). The "resurrect" would be a one-liner "load this pending restore request if it's set". Would we need more?

**Comment by thebnich** [comment]
> There's some questions there about the overhead of creating a webview but not loading anything in it that I really don't know much about.

Yeah -- my guess that an unloaded webView uses inconsequential amount of memory, but that should be easy to verify using the Allocations instrument (just do a before/after memory comparison after initializing 10 empty webViews or something).

**Comment by wesj** [comment]
Well memory isn't even the real consideration here. Its startup speed. I still imagine that might be faster if we can somehow avoid webview process creation (although we use one process for all webviews, so I don't even know that that is the problem)... but creating and attaching the webviews to the browserviewcontroller IS slow.

Even ignoring that I still don't think we gain much. The zombify method isn't used much here, (and wouldn't be necessary if we removed the memory warning stuff, but I think its nice for symmetry). We'd still have a resurrect-type method. We already have the stored url here (we don't have to use it as a boolean either, which means we can evict a webview at any time we want).

We have to store the navDelegate here as well as the wkWebViewConfiguration. The rest of this patch is basically what you describe, plus a lot of ?'s. We have to add a dontLoadUrl(url) method to load a zombie url I guess. We could delay attaching the webview until its selected. At some point its easier to just not create it be really explicit here instead of trying to hide what's happening IMO.

**Comment by thebnich** [comment]
Rather than guessing the cost, I added this little chunk to the bottom of `viewDidLoad`:
```
        let start = NSDate.timeIntervalSinceReferenceDate()
        for i in 0..<10 {
            let webView = WKWebView()
        }
        let stop = NSDate.timeIntervalSinceReferenceDate()
        println("BRN: startup cost: \(stop - start)")
```

That chunk took over 700ms (over 70ms per tab!) just to *create* them -- that's without even adding them to the view hierarchy. So it looks like you're right: the "simple" path won't be good enough here.

**Comment by st3fan** [comment]
Also, when measuring memory usage, remember two things: first, that a large chunk is happening outside of the main application process. You may actually be able to see the external processes that are being created. (On OS X they will have the site name in their process name). Second, the behaviour (of memory in general and webviews specifically) and memory usage is different on the simulator. Be sure to check on a real device.

File: Client/Frontend/Browser/FaviconManager.swift [hunk 1]
```diff
@@ -17,7 +17,7 @@ class FaviconManager : BrowserHelper {
         if let path = NSBundle.mainBundle().pathForResource("Favicons", ofType: "js") {
             if let source = NSString(contentsOfFile: path, encoding: NSUTF8StringEncoding, error: nil) as? String {
                 var userScript = WKUserScript(source: source, injectionTime: WKUserScriptInjectionTime.AtDocumentEnd, forMainFrameOnly: true)
-                browser.webView.configuration.userContentController.addUserScript(userScript)
+                browser.webView?.configuration.userContentController.addUserScript(userScript)
```
Comment (thebnich) [inline Client/Frontend/Browser/FaviconManager.swift]: Maybe worth using `!` instead of `?` in the helpers? Having an optional `webView` opens up a new set of bugs where things might be "skipped" if the `webView` isn't created yet. We should probably strongly enforce that `Browser`s have a valid `webView` for one-off events like this (when helpers are created) -- otherwise, this init does nothing, and the user scripts never get set on the `webView`, and the browser is gimped.

File: Client/Frontend/Browser/Browser.swift [hunk 2]
```diff
@@ -13,30 +13,28 @@ protocol BrowserHelper {
...
-    let webView: WKWebView
+    var webView: WKWebView? = nil
+
     var browserDelegate: BrowserDelegate? = nil
     var bars = [SnackBar]()
     var favicons = [Favicon]()
+
     var screenshot: UIImage?
+    var lastRequest: NSURLRequest? = nil
+    let configuration: WKWebViewConfiguration
```
Comment (thebnich) [inline Client/Frontend/Browser/Browser.swift]: This should probably be `private` to avoid misusing it (see comments below).

File: Client/Frontend/Browser/PasswordHelper.swift [hunk 3]
```diff
@@ -30,7 +30,7 @@ class PasswordHelper: BrowserHelper {
         if let path = NSBundle.mainBundle().pathForResource("PasswordHelper", ofType: "js") {
             if let source = NSString(contentsOfFile: path, encoding: NSUTF8StringEncoding, error: nil) as? String {
                 var userScript = WKUserScript(source: source, injectionTime: WKUserScriptInjectionTime.AtDocumentEnd, forMainFrameOnly: true)
-                browser.webView.configuration.userContentController.addUserScript(userScript)
+                browser.configuration.userContentController.addUserScript(userScript)
```
Comment (thebnich) [inline Client/Frontend/Browser/PasswordHelper.swift]: Careful, you're setting the `configuration` of the `browser`, not the `webView`. As described in the [WKWebView docs](https://developer.apple.com/library/ios/documentation/WebKit/Reference/WKWebView_Ref/#//apple_ref/occ/instm/WKWebView/initWithFrame:configuration:), the `configuration` object is used only to initialize the webView object, after which it's "dead". If the webView is already created by the time this helper is initialized, this line will have no effect.

I recommend making `configuration` private as mentioned above to help us avoid this.

File: Client/Frontend/Browser/TabManager.swift [hunk 4]
```diff
@@ -229,6 +229,14 @@ class TabManager : NSObject {
         let storedTabs: [RemoteTab] = optFilter(tabs.map(Browser.toTab))
         storage?.insertOrUpdateTabs(storedTabs)
     }
+
+    func didReceiveMemoryWarning() {
+        for (index, tab) in enumerate(tabs) {
+            if selectedIndex != index {
+                tab.zombify()
```
Comment (thebnich) [inline Client/Frontend/Browser/TabManager.swift]: I still think we should avoid doing our own zombification, for a few reasons:
 1. The web view already does it for us. Why reinvent the wheel and potentially introduce new bugs for existing functionality?
 1. As I mentioned before, the OS zombifies each web view on an as-needed basis, rather than all at once. If I have 20 tabs open and I get a low memory warning, it's a much better user experience if only a couple of those tabs are closed as opposed to all of them.

File: Client/Frontend/Browser/TabManager.swift [hunk 5]
```diff
@@ -122,6 +120,7 @@ class TabManager : NSObject {
         }
 
         assert(tab === selectedTab, "Expected tab is selected")
+        selectedTab?.resurrect()
```
Comment (thebnich) [inline Client/Frontend/Browser/TabManager.swift]: `resurrect` is a bit of a misnomer -- you're "resurrecting" tabs that are brand new. Maybe `initializeWebView` would be a more appropriate name?

File: Client/Frontend/Browser/TabTrayController.swift [hunk 6]
```diff
@@ -488,7 +489,14 @@ class TabTrayController: UIViewController, UITabBarDelegate, UICollectionViewDel
...
             }
-            cell.background.image = tab.screenshot
+
+            if let img = tab.screenshot {
+                cell.background.image = img
+            } else if let thumbs = profile.thumbnails as? SDWebThumbnails {
+                let key = SDWebThumbnails.getKey(tab.url!)
+                cell.background.moz_getImageFromCache(key, cache: thumbs.cache, completed: { img, err, type, key in
+                })
+            }
```
Comment (thebnich) [inline Client/Frontend/Browser/TabTrayController.swift]: FWIW, @st3fan's recent change in [#491](https://github.com/mozilla/firefox-ios/pull/491/files#diff-221fd685cced62e8c7406c0b5f24ed6cR242) saves the screenshot in the session, so we shouldn't be losing screenshots anymore.

File: Client/Frontend/Browser/TabTrayController.swift [hunk 7]
```diff
@@ -488,7 +489,14 @@ class TabTrayController: UIViewController, UITabBarDelegate, UICollectionViewDel
             } else {
                 cell.favicon.image = UIImage(named: "defaultFavicon")
             }
-            cell.background.image = tab.screenshot
+
+            if let img = tab.screenshot {
+                cell.background.image = img
+            } else if let thumbs = profile.thumbnails as? SDWebThumbnails {
+                let key = SDWebThumbnails.getKey(tab.url!)
```
Comment (thebnich) [inline Client/Frontend/Browser/TabTrayController.swift]: Unsafe, tab URL can be nil.

File: Client/Frontend/Browser/TabTrayController.swift (hunk 6 above)
Comment (wesj) [inline Client/Frontend/Browser/TabTrayController.swift]: Argh. Why did we do that? We already pay for storage for these. Now we're storing them twice?

Comment (thebnich) [inline Client/Frontend/Browser/TabTrayController.swift]: What do you mean? We aren't storing screenshots anywhere else -- just the smaller, poorer quality thumbnails for the top sites panel (AFAIK).

Comment (wesj) [inline Client/Frontend/Browser/TabTrayController.swift]: I forgot those are half size. I'd still rather we just stored them normal size once than this. i.e. one place. one api. etc.

Comment (thebnich) [inline Client/Frontend/Browser/TabTrayController.swift]: Heh, they're actually <1/10th the size, so that could be a big difference if we store a lot of thumbnails (taking the top 3:2 thumbnail from a 9:16 screen is 3/8, then half the width and height would be (3/8)/4, or 3/32).

File: Client/Frontend/Browser/FaviconManager.swift (hunk 1 above)
Comment (wesj) [inline Client/Frontend/Browser/FaviconManager.swift]: Good idea. At one point I had the tabs doing some jumps through hoops to store and restore these, but I just decided it was easier to use the callbacks and let these do it themselves. I'll make this non-optional.

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 8]
```diff
@@ -738,6 +739,43 @@ extension BrowserViewController: BrowserToolbarDelegate {
...
+
+    func browser(browser: Browser, didCreateWebView webView: WKWebView) {
+        webView.scrollView.delegate = self
+        webViewContainer.insertSubview(webView, atIndex: 0)
+        webView.snp_makeConstraints { make in
+            make.edges.equalTo(self.webViewContainer)
+        }
+
+        // Observers that live as long as the tab. Make sure these are all cleared
+        // in didRemoveTab below!
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserViewController.swift]: s/didRemoveTab/willDeleteWebView/

File: Client/Frontend/Browser/TabTrayController.swift [hunk 9]
```diff
@@ -549,7 +550,10 @@ class TabTrayController: UIViewController, UITabBarDelegate, UICollectionViewDel
             } else {
                 cell.favicon.image = UIImage(named: "defaultFavicon")
             }
-            cell.background.image = tab.screenshot
+
+            if let img = tab.screenshot {
+                cell.background.image = img
+            }
```
Comment (thebnich) [inline Client/Frontend/Browser/TabTrayController.swift]: Nit: I think this is an unnecessary change since setting `cell.background.image` to `nil` is the same as not setting it at all.

File: Client/Frontend/Reader/ReadabilityBrowserHelper.swift [hunk 10]
```diff
@@ -23,7 +23,7 @@ class ReadabilityBrowserHelper: BrowserHelper {
            let readabilityBrowserHelperSource = NSMutableString(contentsOfFile: readabilityBrowserHelperPath, encoding: NSUTF8StringEncoding, error: nil) {
             readabilityBrowserHelperSource.replaceOccurrencesOfString("%READABILITYJS%", withString: readabilitySource as String, options: NSStringCompareOptions.LiteralSearch, range: NSMakeRange(0, readabilityBrowserHelperSource.length))
             var userScript = WKUserScript(source: readabilityBrowserHelperSource as String, injectionTime: WKUserScriptInjectionTime.AtDocumentEnd, forMainFrameOnly: true)
-            browser.webView.configuration.userContentController.addUserScript(userScript)
+            browser.webView?.configuration.userContentController.addUserScript(userScript)
```
Comment (thebnich) [inline Client/Frontend/Reader/ReadabilityBrowserHelper.swift]: Should this be `webView!` like the other helpers?

**Comment by thebnich** [comment]
Nice, this looks good to me, though you'll have to do a bit more rebasing after #521 (sorry). That PR adds a `testWebViewDisposed` to make sure the webview gets disposed when the browser is removed, so it'd be good to make sure that still passes with this (especially considering you moved some deregistration into the `Browser` `deinit`).
